### ISSUE TITLE
Does relay matching behavior need to be negotiated in-band?

### ISSUE THREAD

**Issue by afrind** [moq-wg/moq-transport#252]
This came out of discussion on #204

When a relay receives an SUBSCRIBE it may need to issue its own SUBSCRIBE to a publisher that has previously ANNOUNCE'd a namespace matching that track.  The current draft doesn't specify how this match is done.  A relay could implement an exact match on Track Namespace (requires a tweak to SUBSCRIBE), or could do a longest prefix match on track name vs all ANNOUNCE'd namespaces, or possibly something else.  Is it ok to leave this up to implementations or should the draft specify this behavior more exactly?  If more than one matching type is allowed, does it need to be negotiated, and by whom (publisher, subscriber, or both)?

**Comment by afrind** [comment]
An issue that came up in interop that is somewhat related - @kixelated's relay was implemented to take the namespace from the connect URL, because in his use-case, ANNOUNCE can be made redundant with the connect URL (eg it contains a namespace and enough auth).  However, this caused an interop problem because my chat application assumes a certain functionality from relays, and that includes supporting ANNOUNCE messages to determine routing to publishers.  Do we need to be more prescriptive about relay behavior?

**Comment by suhasHere** [comment]
ANNOUNCE is what my implementation expects as well. A producer can announce track namespaces at any point (after being connected).  Overloading connect URL  with track info is limiting

**Comment by fluffy** [comment]
This is going to be a slightly long comment but this is how I think this should work. 

The catalog has two fields for the track namespace , and track name. We also specify in the catalog a way to form a URL from these two fields. See Teds proposal and that is catalog specific.  The things we put in these track namespaces and track name fields are already canonicalized  in a way defined by the catalog and URL. 

When the application using the moq transport puts these two fields into message, they are treated as just a bunch of bits and any comparison of them is done with bitwise compare. A relay that needs to check if a namespace matches another does not need to worry about string prep or escape encoding of the bits, the will all have been done when the catalog was formed. 

This allows the relays to work faster by just doing bitwise compare yet at the same time allows us to use URI at the application level. The transport draft just needs to deal with a bag of bits. The catalog draft pushes the work to the URL definition, and the URL definition does the heavy lifting but that is where this type of stuff should be

**Comment by afrind** [comment]
@fluffy Just to clarify that when you say "bitwise compare" you also mean exact matching the entire namespace field from announce to the one in subscribe?

**Comment by kixelated** [comment]
> The catalog has two fields for the track namespace , and track name.

I would like to decouple the media layer from the transport layer. A tool like ffmpeg should be able to generate a MoQ catalog without knowing the namespace it will be served from, much like a tool like zip should be able to generate a file without knowing the path it will be stored.

HLS/DASH provides this functionality with absolute or relative paths. If the playlist contains relative paths, then it can be served from any HTTP host/path provided the segments are served from the same host/path. If you want to decouple the two (ex. application serves playlist, CDN serves segments), then you use absolute URLs.

To accomplish the same thing, the namespace should be optional in a MoQ catalog. This would mean media tracks are served from the same namespace as the catalog: whatever that might be.

**Comment by kixelated** [comment]
To expand further, I think that MoQ broadcasts should be namespace agnostic: a broadcast ingested with namespace XXX is NOT required to be delivered as XXX.

I understand where Cullen is coming from: an application like WebEx can use the same namespace for contribution/distribution because it controls the full pipeline. The application has all of the information required for all future routing/identification so it can set it on the first hop and expect any relays to blindly proxy.

However, for a more generic or distributed system, this is not true. A client like OBS knows nothing about how the broadcast will eventually be distributed, fanned out, or the requirements of potentially multiple CDNs that might be used, nor will those requirements be the same between vendors (present or future).

So I don't think we should require that input namespace == output namespace. In the HTTP world, this would be like requiring that all HTTP paths match the origin. It's quite restrictive and just means a CDN needs find another way to perform routing, ex. a header/parameter.

I propose we emulate the HTTP model where there's non in-band notification that a path/namespace was modified. An origin might serve `/foo.html` and a CDN might decide to serve it as `/customer/asdgsa13/foo.html`. Any namespace remapping from XXX -> YYY is business logic and not signaled over the wire.

**Comment by afrind** [comment]
> I think that MoQ broadcasts should be namespace agnostic: a broadcast ingested with namespace XXX is NOT required to be delivered as XXX.

I think we have agreement on that from past discussions.  

Can the catalog draft have a mechanism to inherit the namespace for tracks from somewhere else, like a relative URL in an HTML page?

I'm not sure this is relevant for how relays perform matching between SUBSCRIBE and ANNOUNCE though.  The only available information in announce *is the namespace*.

**Comment by kixelated** [comment]
> I'm not sure this is relevant for how relays perform matching between SUBSCRIBE and ANNOUNCE though. The only available information in announce _is the namespace_.

Yeah I'm probably rambling about something unrelated. But in order to support namespace remapping, the downstream ANNOUCE/SUBSCRIBE won't necessarily match the downstream ANNOUCE/SUBSCRIBE.

**Comment by wilaw** [comment]
> To accomplish the same thing, the namespace should be optional in a MoQ catalog. This would mean media tracks are served from the same namespace as the catalog: 

> Can the catalog draft have a mechanism to inherit the namespace for tracks from somewhere else, like a relative URL in an HTML page?

This is a useful property. I opened [#33](https://github.com/moq-wg/catalog-format/issues/33) to address this issue and PR https://github.com/moq-wg/catalog-format/pull/34 to fix it.

**Comment by fluffy** [comment]
So if the input full track name does not equal output full track name ( for some version of matching the two ), what is it that tells the relays there is mapping between the two. I'm not objecting to some other way to do this, I just don't understand what it is that tells the relays to do that.  Keep in mind relays do not need to read the catalog.

**Comment by fluffy** [comment]
To clarify the  "bitwise compare" ... yes I mean exact matching the entire namespace field from announce to the one in subscribe? If the namespace was utf-8 or even a url restricted utf-8, that would match. I am just aruging that a relay should not have to verify every string is a valid utf-8 string, it should just match the bits.

**Comment by kixelated** [comment]
> So if the input full track name does not equal output full track name ( for some version of matching the two ), what is it that tells the relays there is mapping between the two. I'm not objecting to some other way to do this, I just don't understand what it is that tells the relays to do that. Keep in mind relays do not need to read the catalog.

HTTP proxies will not modify the path by default. A downstream request for `/foo` arrives, there's some rules to match it to a backend, and then an upstream request for `/foo` is sent. A simple MoQ relay would work the same way.

However, you can configure a HTTP proxy to perform special behavior, including rewriting the path. Note that a the HTTP server does rewrite the HLS/DASH playlist provided they are using relative URLs since both the playlist and segment would match the same prefix rules.

For example, here's a simple nginx configuration:

```nginx
server {
    # match any requests for `akamai.com`
    server_name akamai.com;

    # match any requests that start with /webex
    location /webex {
        # perform custom auth based on prefix
        auth_webex "$WEBEX_KEY"

        # Route to the webex origin and remove /webex from the path
        # ex. /webex/foobar => https://origin.webex.com/foobar
        proxy_pass "https://origin.webex.com";
    }

    # match any requests that start with /twitch
    location /twitch {
        # ex. /twitch/foobar => https://video-edge-cdn.hls.twitch.tv/foobar
        proxy_pass "https://video-edge-cdn.hls.twitch.tv";
    }
}

# alternatively, match based on SNI
server {
    # route any requests for webex.akamai.com straight to the origin
     server_name webex.akamai.com;

     location / {
       # note the path is not written
       proxy_pass "https://origin.webex.com";
    }
}
```

I would allow the same exact thing for MoQ, but of course use our terminology instead. I think a key improvement with MoQ over HTTP is actually the namespace/name tuple. I propose:

- A relay SHOULD NOT change the track name. That means a producer can make a track called "video_480p" and reference it in a catalog, inheriting the current namespace. 
- A relay MAY change the track namespace when negotiated out-of-band. It's RECOMMENDED that an application support being served from arbitrary namespaces, including using relative namespaces when tracks reference each other.

Basically, the track name is controlled by the application (encoder/decoder) while the track namespace is controlled by the relay (hop-by-hop). This makes a ton of sense if the namespace is primarily meant for routing (critical for relays), and it allows relays to do stuff like encode the route or a unique ID into the namespace itself.

**Comment by kixelated** [comment]
Going back to the original issue Alan posted (if I understand it correctly), I think we need to support out-of-band negotiation given the nature of fanout. Maybe we can provide something in-band but it won't always be possible.

For example, an OBS client sends:
```
CONNECT https://ingest.twitch.tv
ANNOUNCE twitch/kixelated
```

In theory, the server could reply with the information required to watch the broadcast, including any remapping:

```
ANNOUNCE_OK connect=https://twitch.akamai.com namespace=kixelated_123
```

But this gets really messy when you factor in auth, multiple CDNs, multiple edge hosts, and just the nature of ingest being on the other side of the pipe from distribution

**Comment by afrind** [comment]
Individual Comment: 

> So if the input full track name does not equal output full track name ( for some version of matching the two ), what is it that tells the relays there is mapping between the two

My opinion is that anything that does this kind of rewriting is out-of-band, at least for now?  For a completely generic relay input=output.

**Comment by wilaw** [comment]
> Can the catalog draft have a mechanism to inherit the namespace for tracks from somewhere else, like a relative URL in an HTML page?

Yes. I opened issue https://github.com/moq-wg/catalog-format/issues/33 to address this and PR https://github.com/moq-wg/catalog-format/pull/34 to fix it. I really like the flexibility that relative names bring. We know this works at scale for current web and media delivery across disparate systems. I'd hope that relative track names are the normal mode of operation. We should allow absolute namespaces in catalogs to allow mixing between sources, however I think this will be an edge case. 

By default, a relay SHOULD NOT modify the track namespace or track name. However, what a distribution network does within the bounds of its own network is up to it. As it restores the original namespace and name at its boundaries (to match the catalog), then external systems cannot tell and distribution intent is preserved. Additionally, there will be business logic and distribution agreements between CDNs, coordinated by the content distributor, to rewrite namespaces between networks. Such namespace rewriting is common between networks and should not be prohibited. 

Here's how I envisage a scalable solution working. When a twitch user wants to use OBS to publish, they would log in to Twitch.com and authenticate against the Twitch CMS. It would vend them two URLS:

a connect  and publish URL `https://ingest.akamai.twitch.tv/twitch/kixelated`
a playback URL to share with their followers `https://playback.fastly.twitch.tv/premiergaming/gamer34/catalog. `

The OBS client would use this to 

```
CONNECT https://ingest.akamai.twitch.tv/
ANNOUNCE twitch/kixelated

```
The client then publishes twitch/kixelated/catalog, a track which describes two relative tracks - audio and video. 

```
video
audio
```

The playback client connects to https://playback.fastly.twitch.tv and subscribes to premiergaming/gamer34/catalog. After parsing the catalog it subscribes  to

```
premiergaming/gamer34/video
premiergaming/gamer34/video
```

The Fastly relays have been configured by Twitch, out-of--band, so that when they go forward to origin (which is Akamai), they do a namespace rewrite to

```
premiergaming/gamer34/video -> twitch/kixelated/video
premiergaming/gamer34/audio -> twitch/kixelated/audio
```

The subscriptions  make their way back to the OBS instance. The encoder prepare Moqt Objects with `twitch/kixelated/video ` in their headers. As these are received by Fastly, business logic rewrites them to `premiergaming/gamer34/video`, caches them as such and they are delivered to the client.

**Comment by suhasHere** [comment]
As per the discussions in person,  it is more nuanced than renaming few bits in the header.  It might break e2e encryption /auth for example or reencryption in some cases. As this is very application specific,  one needs to be take into consideration the costs as it may not be just rewrite few bits in and out of network edges.  The more appropriate or clearer framing may be -  whenever the name changes, it is ,in a way, republishing the content with a new name ..

**Comment by suhasHere** [comment]
> As it restores the original namespace and name at its boundaries (to match the catalog)

just to note, the network wouldn't have access to the catalog

**Comment by ianswett** [comment]
Conclusion is that we'll use exact matching like specified in #329
